include/hlslib/xilinx/Simulation.h

```diff
@@ -62,13 +69,43 @@ class _Dataflow {
  public:
   inline static _Dataflow& Get() { // Singleton pattern
     static _Dataflow df;
-    return df; 
+    return df;
   }
 
   template <class Ret, typename... Args>
   void AddFunction(Ret (*func)(Args...), non_deducible_t<Args>... args) {
-    threads_.emplace_back(func, passed_by(std::forward<Args>(args),
-                                          std::is_reference<Args>{})...);
+    if (subflow_active_) {
+      subflows_.back().AddFunction(func, args...);
+    } else {
+      threads_.emplace_back(func, passed_by(std::forward<Args>(args),
+                                            std::is_reference<Args>{})...);
+    }
+  }
+
+  template <class Ret, typename... Args>
+  void AddSubflow(Ret (*func)(Args...), non_deducible_t<Args>... args) {
+    if (subflow_active_) {
+      subflows_.back().AddSubflow(func, args...);
+    } else {
+      subflows_.emplace();
+      std::unique_lock<std::mutex> lock(mutex_);
+      subflow_active_ = true;
+      threads_.emplace_back(func, passed_by(std::forward<Args>(args),
+                                            std::is_reference<Args>{})...);
+      while (subflow_active_) cond_var_.wait(lock);
+    }
+  }
+
+  inline void FinalizeSubflow() {
+    if (subflows_.back().subflow_active_) {
+      subflows_.back().FinalizeSubflow();
+    } else {
+      std::unique_lock<std::mutex> send_lock(mutex_);
+      subflow_active_ = false;
+      send_lock.unlock();
+      cond_var_.notify_one();
```

Comment:
```suggestion
      cond_var_.notify_one();
      send_lock.unlock();
```

The idiomatic pattern is to notify before unlocking, so that you maintain the ordering of other threads waiting on the lock (not that it really matters for our use case).

Reply:
Interesting, I read the opposite somewhere...

Reply:
Indeed not relevant to our use case, but I'm always in favor of following idiomatic patterns.

Reply:
I did some reading up on this, and I believe you're right: I was thinking that we want to maintain thread order on the lock, but this is bogus. Let's keep it the way you had it